Support operator with no input

cw75 opened the pull request:
## Describe your changes and why you are making these changes
One way to enable this is to make module-level variables for the `dag` and `api_client` objects. The following change seems to work.

## Related issue number (if any)

## Checklist before requesting a review
- [x] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [x] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [x] If this is a new feature, I have added unit tests and integration tests.
- [x] I have run the integration tests locally and they are passing.
- [x] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [x] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

kenxu95 commented on sdk/aqueduct/aqueduct_client.py:
```diff
@@ -88,10 +90,12 @@ def __init__(
             str, IntegrationInfo
         ] = self._api_client.list_integrations()
         self._dag = DAG(metadata=Metadata())
+        dag.__GLOBAL_DAG__ = self._dag
```
the global api client is fine with me. But what happens if we user constructs multiple aqueduct clients? Do we want to wipe out the previous dag completely? I'd rather us not overwrite the global dag if it already exists, so we'd have two clients using the same underlying dag, which is weird but seems less destructive.

kenxu95 commented on sdk/aqueduct/decorator.py:
```diff
@@ -71,17 +73,19 @@ def wrap_spec(
+        api_client = api_client_module.__GLOBAL_API_CLIENT__
+    else:
+        dag = input_artifacts[0]._dag
+        api_client = input_artifacts[0]._api_client
```
I'm not sure we should even keep the else statement once we have global variables.

kenxu95 commented on sdk/aqueduct/decorator.py:
```diff
@@ -71,17 +73,19 @@ def wrap_spec(
+    if len(input_artifacts) == 0:
+        print("This operator takes no artifact input.")
+        dag = dag_module.__GLOBAL_DAG__
+        api_client = api_client_module.__GLOBAL_API_CLIENT__
```
There's another assumption we're sort of breaking here. Previously, all starting operators have to be constructed through the aqueduct client, so the client would have to be instantiated before running operators. Now, a user may expect their decorated function to run even without the client instantiated, and will be confused as to why the client needs to be constructed even if it's not directly used.

I don't think we can avoid this dependency right now, but what if we communicated this better by:
- initialize the global dag to `DAG(metadata=Metadata())` instead of None.
- initialize the global api client without the api_key and server_address. We can add a `configure(api_key, server_address)` method to set these fields up. If we attempt to use the api_client without configuring it, the user will get a very informative error about how they need to set the api_key and server_address first, currently by instantiating the client. Later on, when we have the concept of a `aq.config(...)` global config object, we can configure the api_client directly there.

cw75 commented:
@kenxu95 addressed comments. I had to add dag variable reset logic to integration tests for some of them to pass. I think we had an assumption that when the client gets destroyed the dag var will be reset, which is no longer true, so I had to manually reset. I noticed that when I ran the integration test locally, "test_check_failure_with_varying_severity" is failing with a preview error, but it's not showing here. Not sure why.

cw75 commented:
@kenxu95 added an integration test.

kenxu95 commented on sdk/aqueduct/api_client.py:
```diff
@@ -107,14 +107,17 @@ class APIClient:
         if self.aqueduct_address.endswith("/"):
             self.aqueduct_address = self.aqueduct_address[:-1]
 
+        self.configured = True
```
Should configured be set at the end of this function? If we fail to validate the api_client, should we allow it to be "configured"?

We can get rid of the logic below, since that is actually solved by the constructor not taking in any arguments.
```
# If a dummy client is initialized, don't perform validation.
        if self.api_key == "" and self.aqueduct_address == "":
```

kenxu95 commented on sdk/aqueduct/api_client.py:
```diff
@@ -132,13 +135,22 @@ def __init__(self, api_key: str, aqueduct_address: str):
+            raise Exception(
+                "API client has not been configured, please complete the configuration \
+                by initializing an Aqueduct client with the api key and the server address."
+            )
```
Instead of having to label every single helper in this class, I think all we need to check configs for is:
`construct_base_url()`, `construct_full_url()`, `url_prefix()`, and `generate_auth_headers()`. The reason I think this makes more sense is because each of these helpers directly uses the methods we've set in `configure()`.

It would also be great to move `generate_auth_headers()` into the APIClient, I'm not sure why these were in utils to begin with.

kenxu95 commented on sdk/aqueduct/tests/metric_test.py:
```diff
@@ -28,14 +29,13 @@ def test_metric():
     artifact_id = generate_uuid()
     artifact_name = "artifact"
-    api_client = APIClient("", "")
+    api_client.__GLOBAL_API_CLIENT__.configure("", "")
```
I don't think we need this line anymore right.

kenxu95 commented on sdk/aqueduct/tests/utils.py:
```diff
@@ -149,8 +147,8 @@ def default_table_artifact(
         outputs=[artifact_id],
     )
-    dag = _construct_dag(
+    dag_module.__GLOBAL_DAG__ = _construct_dag(
```
hmm because `_construct_operator()` is used by multiple test cases, I wonder if its safer to just use the local dag for the artifacts. These unit tests are not supposed to use decorators or the aqueduct client methods anyways, since its supposed to validate lower-level constraints.

kenxu95 reviewed (changes requested):
I have a few questions/suggestions but other than that it looks really good!

cw75 commented:
@kenxu95 addressed all comments.

kenxu95 commented on sdk/aqueduct/api_client.py:
```diff
@@ -166,6 +186,7 @@ def _test_connection_protocol(self, try_http: bool, try_https: bool) -> bool:
                 self._test_url(url)
                 return False
             except Exception as e:
+                print(e)
```
I think we can probably just do `logger.error()` here, which already includes the exception.
